doc/specs/#476 - Pinned packages.md

```diff
@@ -0,0 +1,180 @@
+
+With the new `--include-pinned` argument, Upgrade should show a separate table in addition to the regular table which shows possible upgrades to pinned packages.
+
+```
+<...>
+Upgrades to pinned packages: 
+
+Name                                                 Id                                Version       Available     Source
+-------------------------------------------------------------------------------------------------------------------------
+Clippy for Azure 2011 Datacenter                     Microsoft.Clippy.2011.Datacenter  2.35.0        2.35.1.2      winget
+```
+
+
+```
+winget <upgrade/install> --include-pinned <package>
+```
+
+Upgrade and Install will now need to use the same argument to bypass the pin for a package temporarily, in which case the pin is updated to pin the new version after the installation completes.
+
+```
+winget install --pin <package>
+```
+
+Install should also be extended to allow the user to pin a package at install time.
+
+
+
+
+In addition to commands, there should be a new `Pinned` key in exported package lists, allowing users to move their pins to different systems:
+
+```
+// ...
+    {
+      "Packages": 
+      [
+        {
+          "PackageIdentifier": "Microsoft.Bob",
+          "PackageVersion" : 2.35.0,
+          "Pinned" : "true"
+        },
+      ],
+    }
+// ...
+```
+
+Lastly, there should be good error messaging throughout the Windows Package Manager in order to explain to users when something doesn't go quite right. 
+
+As an example, consider the situation where a user has pinned their favorite package, Microsoft Bob:
+
+```
+PS C:\Users\billg> winget install --pin Microsoft.Bob
+Found Microsoft Bob [Microsoft.Bob] Version 2.32.0
+This application is licensed to you by its owner.
+Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
+Downloading https://aka.ms/importantsystemdependencies/bob.exe
+██████████████████████████████ 2.08 MB / 2.08 MB
+Successfully verified installer hash
+Starting package install...
+Successfully installed
+Microsoft.Bob has been pinned at version 2.32.0.
+```
+
+Then, later, that user tries to install a package that is dependent on a newer version of Microsoft Bob. The Windows Package Manager should not upgrade Microsoft Bob automatically, but instead should notify the user and give them information on how to continue:
+
+```
+PS C:\Users\billg> winget install --pin Microsoft.RoverTheDog
+Found Rover the Dog [Microsoft.RoverTheDog] Version 5.1
+This application is licensed to you by its owner.
+Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
+This package requires the following dependencies:
+  - Packages
+      Microsoft.Bob, version 2.39.0.
+
+Microsoft Bob is currently pinned at version 2.32.0, but Rover the Dog 5.1 requires at least version 2.39.0. 
+To upgrade Microsoft Bob and other pinned dependencies, add the argument --include-pinned to your previous command.
```

Comment:
I've gone back and forth between just "informing" users and giving an interactive [Y/N] sort of dialog to see if the user wants to perform suggested actions.

Reply:
A [Y/N] prompt would be more user-friendly, but it would make it harder to use winget in automation scenarios. Is there a way on Windows to detect if a user is interactively using a shell vs a script? I know the way to tell apt on linux not to ask questions is via an environment variable, but that seems kludgy.

Reply:
Any of the [Y/N] style prompts are likely suitable for a setting to specify a default.

Reply:
> I know the way to tell apt on linux not to ask questions is via an environment variable, but that seems kludgy.

Sorry for being a little off topic. Just want to make sure you're talking about `export DEBIAN_FRONTEND=noninteractive`

Reply:
I've requested that @lechacon do more extensive interactivity controls for another review.  The goal would be to have a common option that disables interactivity, as well as a setting.  Interactivity in general is not a good way to handle these cases, because the COM (and thus PowerShell eventually) entrypoints cannot use that pattern.  But with a more generic and extensive set of controls and thoughts about non-interactive defaults, we can use them more (within reason).